client/Components/CodeEditor/CodeEditor.js

```diff
@@ -0,0 +1,108 @@
+import React, { Component } from 'react'
+import axios from 'axios'
+import Select from '@material-ui/core/Select'
+import OutlinedInput from '@material-ui/core/OutlinedInput'
+import MenuItem from '@material-ui/core/MenuItem'
+import configEditor from './EditorConfigFunction'
+
+class AceEditor extends Component {
+  constructor() {
+    super()
+    this.state = {
+      code: '',
+      codeResponse: '',
+      codeError: '',
+      editor: {},
+      editorTheme: 'monokai'
+    }
+    this.configEditorBinded = configEditor.bind(this)
+  }
+
+  componentDidMount() {
+    this.setState({ editor: ace.edit('ace') }, function() {
+      const { editor } = this.state
+      const editorSession = editor.getSession()
+      this.configEditorBinded(editor, editorSession, this.state.editorTheme)
+    })
+  }
+
+  runCode = () => {
+    this.setState({ codeResponse: '', codeError: '' }, () => {
+      return axios
+        .post('/api/code/runcode', { code: this.state.code })
+        .then(({ data }) => {
+          this.setState({
+            codeResponse: String(data)
+          })
+        })
+        .catch(({ response: { data } }) => {
+          this.setState({ codeError: data })
+        })
+    })
+  }
+
+  submit = () => {
```

Comment:
I'm not totally clear on this - it looks like this isn't running anything, but just reinitializing the ace-session and the state.

Reply:
Indeed doing that but before that on line 46, see the axios post call. That call is saivng the code ti the database

Reply:
ah ok. b/c the code needs to be run, AND saved in the db. So both the runCode and submit methods are called when the user clicks the button